chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -157,81 +157,118 @@ pub struct TransactionVerifierDelta {
     token_issuance_cache: ConsumedTokenIssuanceCache,
 }
 
+/// [`TxIndexCache`] that can be enabled or disabled (using a config).
+pub struct GuardedTxIndexCache {
+    enabled: bool,
+    inner: TxIndexCache,
+}
+
+impl GuardedTxIndexCache {
+    fn new(enabled: bool) -> Self {
+        let inner = TxIndexCache::new();
+        Self { enabled, inner }
+    }
+
+    fn from_config(config: &TransactionVerifierConfig) -> Self {
+        Self::new(config.tx_index_enabled)
+    }
+
+    #[cfg(test)]
+    fn new_for_test(map: BTreeMap<OutPointSourceId, CachedInputsOperation>) -> Self {
+        let inner = TxIndexCache::new_for_test(map);
+        let enabled = true;
+        Self { enabled, inner }
+    }
+
+    fn as_ref(&self) -> Option<&TxIndexCache> {
+        self.enabled.then_some(&self.inner)
+    }
+
+    fn as_mut(&mut self) -> Option<&mut TxIndexCache> {
+        self.enabled.then_some(&mut self.inner)
+    }
+
+    /// Take the inner cache, even if disabled
+    fn take_always(self) -> TxIndexCache {
+        self.inner
+    }
+}
+
 /// The tool used to verify transaction and cache their updated states in memory
-pub struct TransactionVerifier<'a, S, U> {
-    chain_config: &'a ChainConfig,
-    storage_ref: &'a S,
-    verifier_config: TransactionVerifierConfig,
-    tx_index_cache: TxIndexCache,
+pub struct TransactionVerifier<CC, S, U> {
+    chain_config: CC,
+    storage: S,
+    tx_index_cache: GuardedTxIndexCache,
     utxo_cache: UtxosCache<U>,
     utxo_block_undo: BTreeMap<TransactionSource, BlockUndoEntry>,
     token_issuance_cache: TokenIssuanceCache,
     best_block: Id<GenBlock>,
 }
 
-impl<'a, S: TransactionVerifierStorageRef> TransactionVerifier<'a, S, UtxosDB<&'a S>> {
-    pub fn new(
-        storage_ref: &'a S,
-        chain_config: &'a ChainConfig,
-        verifier_config: TransactionVerifierConfig,
-    ) -> Self {
+impl<CC, S: TransactionVerifierStorageRef + Clone> TransactionVerifier<CC, S, UtxosDB<S>> {
```

Comment:
I am a bit uncomfortable with the `Clone` here. The idea is that storage is some form of reference, smart pointer or handle and cloning it does not clone the contents. Like for example when cloning a `&T` or using `Arc::clone`.

Maybe we can introduce a marker `trait ShallowClone: Clone {}` that is only implemented for types that do shallow clone. Then use it here instead of plain `Clone` to constrain it a bit more precisely.

Comments / counterproposals?

Reply:
I think the idea with `ShallowClone` is worth implementing

Reply:
> I am a bit uncomfortable with the Clone here. The idea is that storage is some form of reference, smart pointer or handle and cloning it does not clone the contents.

Should it be renamed something like StorageHandle? That way, you know it's a handle to the underlying, just like a database handle.

Reply:
I don't mind creating a ShallowClone trait... at least now I don't think of a reason why we shouldn't do that.

Reply:
> > I am a bit uncomfortable with the Clone here. The idea is that storage is some form of reference, smart pointer or handle and cloning it does not clone the contents.
> 
> Should it be renamed something like StorageHandle? That way, you know it's a handle to the underlying, just like a database handle.

Hmm maybe the field, which I have renamed from `storage` to `storage_ref` should have stayed called `storage_ref`.